### PR TITLE
Cold-start DDP finetuning: concurrent dataset downloads race across ranks. 

### PR DESCRIPTION
**Pull request by anuragg1209**
## Issue
Please link the corresponding GitHub issue. If an issue does not already exist,
please open one to describe the bug or feature request before creating a pull request.

This allows us to discuss the proposal and helps avoid unnecessary work.

## Motivation and Context

When running the finetuning examples with multiple GPUs (torchrun --nproc-per-node=N examples/finetune_classifier.py / finetune_regressor.py) for the first time on a machine with a cold sklearn dataset cache, all N ranks call the sklearn dataset fetcher simultaneously, which can crash the run.

## Public API Changes

-   [x] No Public API changes
-   [ ] Yes, Public API changes (Details below)

---

## How Has This Been Tested?

- Added a public` tabpfn.finetuning.main_process_first()` context manager (in `finetuned_base.py`, reusing `_init_distributed_if_needed` so the NCCL backend/timeout stay defined in one place): rank 0 runs the block (downloads and warms the cache) while other ranks wait at a barrier, then read from the warm cache. Leaves the process group initialized for `fit()` to reuse; no-op in single-process runs.
- Both finetune examples now wrap their dataset fetch in with `main_process_first()`:.

Verified with a 2-rank torchrun test asserting rank 1 only enters the block after rank 0 completes it.

## Checklist

-   [x] The changes have been tested locally.
-   [x] Documentation has been updated (if the public API or usage changes).
-   [x] A changelog entry has been added (see `changelog/README.md`), or "no changelog needed" label requested.
-   [x] The code follows the project's style guidelines.
-   [x] I have considered the impact of these changes on the public API.

---

### CODE REVIEW

File: src/tabpfn/finetuning/finetuned_base.py [hunk 1]
```diff
@@ -128,6 +131,35 @@ def _maybe_setup_ddp(
...
+    leaves it initialized so that a subsequent ``fit()`` reuses it. Call
+    ``torch.distributed.destroy_process_group()`` at the end of your script.
+    No-op when running with a single process.
+    """
+    if int(os.environ.get("WORLD_SIZE", "1")) <= 1:
+        yield
+        return
+
+    _init_distributed_if_needed("cuda")
+    is_main_process = dist.get_rank() == 0
```
[the remark this answers is not in the document]
Comment (anuragg1209) [reply src/tabpfn/finetuning/finetuned_base.py:158]: Fixed.

File: src/tabpfn/finetuning/finetuned_base.py [hunk 2]
```diff
@@ -55,6 +56,8 @@
 logger = logging.getLogger(__name__)
 
 if TYPE_CHECKING:
+    from collections.abc import Iterator
+
     from tabpfn.constants import ModelVersion, XType, YType
```
[the remark this answers is not in the document]
Comment (anuragg1209) [reply src/tabpfn/finetuning/finetuned_base.py:60]: Addressed.

File: src/tabpfn/finetuning/finetuned_base.py [hunk 3]
```diff
@@ -128,6 +130,41 @@ def _maybe_setup_ddp(
...
+        return
+
+    using_ddp, _, _ = _init_distributed_if_needed("cuda")
+    if not using_ddp:
+        # WORLD_SIZE was set by something other than torchrun (no LOCAL_RANK),
+        # so there is no process group to coordinate through.
+        yield
+        return
+
+    is_main_process = dist.get_rank() == 0
```
Comment (psinger-prior) [inline src/tabpfn/finetuning/finetuned_base.py:158]: would we need to check for local rank here? although finetuning is mostly single node anyways

**Review by psinger-prior — APPROVED** [review]
I did not test it (which Im sure you did), but code lgtm

File: src/tabpfn/finetuning/finetuned_base.py (hunk 3 above)
Comment (anuragg1209) [reply src/tabpfn/finetuning/finetuned_base.py:158]: Good catch. For the current setup, they're equivalent. Since multi-node isn't supported here yet, I'd keep the global-rank check for now; happy to change if you'd prefer to future-proof?

Comment (anuragg1209) [reply src/tabpfn/finetuning/finetuned_base.py:158]: Actually,  since multi-node isn't supported by `_maybe_setup_ddp` anyway, this parameter is dormant until that changes. It's future-proofing plus documentation of the assumption. I prefer to move with the current setup for now and merge the branch.

**Review by psinger-prior — APPROVED** [review]
